politeiawww/dcc.go

```diff
@@ -970,3 +1030,262 @@ func dccStatusInSlice(arr []cms.DCCStatusT, status cms.DCCStatusT) bool {
 
 	return false
 }
+
+func (p *politeiawww) processVoteDCC(vd cms.VoteDCC, u *user.User) (*cms.VoteDCCReply, error) {
+	log.Tracef("processVoteDCC: %v", u.PublicKey())
+
+	if vd.Vote != "yes" && vd.Vote != "no" {
+		return nil, www.UserError{
+			ErrorCode:    cms.ErrorStatusInvalidSupportOppose,
+			ErrorContext: []string{"vote string not yes or no"},
+		}
+	}
+	vdr, err := p.cmsVoteDetails(vd.Token)
+	if err != nil {
+		return nil, err
+	}
+	dcc, err := p.getDCC(vd.Token)
+	if err != nil {
+		if err == cache.ErrRecordNotFound {
+			err = www.UserError{
+				ErrorCode: cms.ErrorStatusDCCNotFound,
+			}
+		}
+		return nil, err
+	}
+
+	// Only allow voting on Debated DCC proposals
+	if dcc.Status != cms.DCCStatusDebate {
+		return nil, www.UserError{
+			ErrorCode: cms.ErrorStatusInvalidDCCVoteStatus,
+		}
+	}
+
+	endHeight, err := strconv.ParseUint(vdr.StartVoteReply.EndHeight, 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	bb, err := p.getBestBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check to make sure that the Vote hasn't ended yet.
+	if endHeight < bb {
+		return nil, www.UserError{
+			ErrorCode: cms.ErrorStatusDCCVoteEnded,
+		}
+	}
+
+	challenge, err := util.Random(pd.ChallengeSize)
+	if err != nil {
+		return nil, err
+	}
+
+	vote := convertCastVoteFromCMS(vd)
+	for _, voteOption := range vdr.StartVote.Vote.Options {
+		if voteOption.Id == vd.Vote {
+			vote.VoteBit = strconv.FormatUint(voteOption.Bits, 16)
+			break
+		}
+	}
+	payload, err := cmsplugin.EncodeCastVote(vote)
+	if err != nil {
+		return nil, err
+	}
+	pc := pd.PluginCommand{
+		Challenge: hex.EncodeToString(challenge),
+		ID:        cmsplugin.ID,
+		Command:   cmsplugin.CmdCastVote,
+		CommandID: cmsplugin.CmdCastVote,
+		Payload:   string(payload),
+	}
+
+	responseBody, err := p.makeRequest(http.MethodPost,
+		pd.PluginCommandRoute, pc)
+	if err != nil {
+		return nil, err
+	}
+
+	var reply pd.PluginCommandReply
+	err = json.Unmarshal(responseBody, &reply)
+	if err != nil {
+		return nil, fmt.Errorf("Could not unmarshal "+
+			"PluginCommandReply: %v", err)
+	}
+
+	// Verify the challenge.
+	err = util.VerifyChallenge(p.cfg.Identity, challenge, reply.Response)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode plugin reply
+	_, err = cmsplugin.DecodeCastVoteReply([]byte(reply.Payload))
+	if err != nil {
+		return nil, err
+	}
+
+	return &cms.VoteDCCReply{}, nil
+}
+
+func (p *politeiawww) debateDCC(token, signature string, u *user.User) error {
```

Comment:
Is starting a DCC debate the same as starting a DCC vote?

Reply:
That's correct.  Would you like me to rename Debate to something more akin to starting a vote?

Reply:
I think its easier to follow if consistent terminology is used throughout.  Doesn't really matter which one as long as its consistent.